### PR TITLE
Collect pgstat from QE to enable auto-ANALYZE on partition leaf table.

### PR DESCRIPTION
**Pull request by JunfengYang**
Collect tuple relead pgstat table info from segments. Then the
auto-analyze could consider partition tables now. Since before, we don't
have accurate pgstat for partition table. This kind of info is counted
through the access method on segments and we used to collect them by the
estate es_processed count on QD. So if insert into the root partition
table, we can not know how many tuples inserted into a leaf, autovacuum
never trigger auto-ANALYZE for leaf table.

The idea is, for writer QE, report current nest level xact tables pgstat
to QD through libpq at the end of a query statement. For a single
statement, it wouldn't operate too many tables, so the effort is really
small.
And on QD, retrieve and combine these tables' stat from the dispatch
result and add to current nest level xact pgstats.
Now we can remove the old pgstat collection code on QD.

The pgstat for a table could be view by query `pg_stat_all_tables_inernal`.
And now, except for the scan related counters, other counters should be accurate.
On master, the table's pgstat of scan related counters are not gathered
from segments yet, this requires extra work. The current implementation
is already enough for supporting auto-ANALYZE on partition table.

Remove `auto_stats` should be evaluated after this PR. Since auto-ANALYZE
seems to be a better solution.
#### TODO
- [x] Add test case for auto-ANALYZE on partition table.
- [x] Add test case for pgstats collected on QD.
- [x] Evaluate performance impact.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by JunfengYang** [comment]
Run pgbench to compare the performance impact on master and this PR.
The result shows no performance decline.

Result on master branch:
```
gpconfig -c gp_enable_global_deadlock_detector -v on
gpconfig -c log_statement -v none

transaction type: <builtin: TPC-B (sort of)>
scaling factor: 500
query mode: simple
number of clients: 30
number of threads: 30
duration: 300 s
number of transactions actually processed: 290661
latency average = 30.965 ms
latency stddev = 13.261 ms
tps = 968.647536 (including connections establishing)
tps = 968.681226 (excluding connections establishing)
statement latencies in milliseconds:
         0.004  \set aid random(1, 100000 * :scale)
         0.001  \set bid random(1, 1 * :scale)
         0.001  \set tid random(1, 10 * :scale)
         0.001  \set delta random(-5000, 5000)
         0.213  BEGIN;
         6.531  UPDATE pgbench_accounts SET abalance = abalance + :delta WHERE aid = :aid;
         4.198  SELECT abalance FROM pgbench_accounts WHERE aid = :aid;
         2.626  UPDATE pgbench_tellers SET tbalance = tbalance + :delta WHERE tid = :tid;
         3.265  UPDATE pgbench_branches SET bbalance = bbalance + :delta WHERE bid = :bid;
         5.941  INSERT INTO pgbench_history (tid, bid, aid, delta, mtime) VALUES (:tid, :bid, :aid, :delta, CURRENT_TIMESTAMP);
         8.184  END;
```

The result on this PR:
```
gpconfig -c gp_enable_global_deadlock_detector -v on
gpconfig -c log_statement -v none

transaction type: <builtin: TPC-B (sort of)>
scaling factor: 500
query mode: simple
number of clients: 30
number of threads: 30
duration: 300 s
number of transactions actually processed: 321493
latency average = 27.995 ms
latency stddev = 11.797 ms
tps = 1071.377491 (including connections establishing)
tps = 1071.416806 (excluding connections establishing)
statement latencies in milliseconds:
         0.004  \set aid random(1, 100000 * :scale)
         0.001  \set bid random(1, 1 * :scale)
         0.001  \set tid random(1, 10 * :scale)
         0.001  \set delta random(-5000, 5000)
         0.200  BEGIN;
         5.065  UPDATE pgbench_accounts SET abalance = abalance + :delta WHERE aid = :aid;
         3.803  SELECT abalance FROM pgbench_accounts WHERE aid = :aid;
         2.485  UPDATE pgbench_tellers SET tbalance = tbalance + :delta WHERE tid = :tid;
         3.129  UPDATE pgbench_branches SET bbalance = bbalance + :delta WHERE bid = :bid;
         5.684  INSERT INTO pgbench_history (tid, bid, aid, delta, mtime) VALUES (:tid, :bid, :aid, :delta, CURRENT_TIMESTAMP);
         7.623  END;
```

**Comment by JunfengYang** [comment]
The reason why autovacuum not consider the mid-layer/root of the partition table is the pgstat for the mid-layer is not collected.
But now, the PG upstream is trying to enable autovacuum(auto-ANALYZE) on the middle layer and root of the partition table.
https://www.postgresql.org/message-id/CAKkQ508_PwVgwJyBY=0Lmkz90j8CmWNPUxgHvCUwGhMrouz6UA@mail.gmail.com

File: src/backend/postmaster/pgstat.c [hunk 1]
```diff
@@ -4529,6 +4539,222 @@ pgstat_send_bgwriter(void)
...
+			records[i].table_stat.t_id, records[i].table_stat.t_shared);
+
+		if (pgstat_info->trans == NULL ||
+			pgstat_info->trans->nest_level != nest_level)
+			add_tabstat_xact_level(pgstat_info, nest_level);
+		trans = pgstat_info->trans;
+		if (bms_is_member(records[i].table_stat.t_id, *oidMap))
+		{
+			/*
+			* Same table pgstat from different QE;
```
Comment (zhangh43) [inline src/backend/postmaster/pgstat.c:4673]: format of comment

File: src/backend/postmaster/pgstat.c [hunk 2]
```diff
@@ -4529,6 +4539,222 @@ pgstat_send_bgwriter(void)
...
+							"Inserted: %ld, updated: %ld, deleted: %ld.",
+							nest_level, tabstat->t_id,
+							trans->tuples_inserted, trans->tuples_updated,
+							trans->tuples_deleted)));
+		}
+
+		if (buf.len > 0)
+		{
+			Assert(stat_data.len > 0);
+			/* Don't mark the result ready when receive this message */
```
Comment (zhangh43) [inline src/backend/postmaster/pgstat.c:4607]: Better to have more comments on the reason why we introduce this ready flag.

**Review by zhangh43 — APPROVED** [review]

File: src/backend/postmaster/pgstat.c [hunk 3]
```diff
@@ -4529,6 +4539,228 @@ pgstat_send_bgwriter(void)
...
+				&stat_data, (char *)&record, sizeof(PgStatTabRecordFromQE));
+			ereport(DEBUG3,
+					(errmsg("Send pgstat for current xact nest_level: %d, table oid: %d. "
+							"Inserted: %ld, updated: %ld, deleted: %ld.",
+							nest_level, tabstat->t_id,
+							trans->tuples_inserted, trans->tuples_updated,
+							trans->tuples_deleted)));
+		}
+
+		if (buf.len > 0)
```
Comment (hidva) [inline src/backend/postmaster/pgstat.c:4604]: `buf` will be uninitialized if trans is NULL. But it looks like that trans will never be NULL=.=

Comment (JunfengYang) [reply src/backend/postmaster/pgstat.c:4604]: @hidva 
Good cache! Actually, I already refactor this part in my new PR https://github.com/greenplum-db/gpdb/pull/11142/commits/97b6b7cca8fb0ee3f910d6fb673f896e1d61ffc0#diff-6161d3e2403ae3e7c0e8434321221ec00f45b6e924eb0825651cb3352a882eaaR4566-R4634

**Pull request merged**